// @umijs/bundler-utoopack/dist/config.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/config.ts
var config_exports = {};
__export(config_exports, {
  getDevUtooPackConfig: () => getDevUtooPackConfig,
  getProdUtooPackConfig: () => getProdUtooPackConfig,
  mergeExtraPostcssPlugins: () => mergeExtraPostcssPlugins
});
module.exports = __toCommonJS(config_exports);
var import_bundler_webpack = require("@umijs/bundler-webpack");
var import_utils = require("@umijs/utils");
var import_pack = require("@utoo/pack");
var import_fs = __toESM(require("fs"));
var import_path = require("path");
function normalizeDefineValue(val) {
  if (!import_utils.lodash.isPlainObject(val)) {
    return JSON.stringify(val);
  } else {
    return Object.keys(val).reduce((obj, key) => {
      obj[key] = normalizeDefineValue(val[key]);
      return obj;
    }, {});
  }
}
function getModularizeImports(extraBabelPlugins) {
  return extraBabelPlugins.filter((p) => /^import$|babel-plugin-import/.test(p[0])).reduce(
    (acc, [_, v]) => {
      const {
        libraryName,
        libraryDirectory,
        style,
        camel2DashComponentName,
        transformToDefaultImport,
        ...rest
      } = v;
      if (Object.keys(rest).length > 0) {
        throw new Error(
          `babel-plugin-import options ${Object.keys(
            rest
          )} is not supported in utoopack bundler`
        );
      }
      if (typeof style === "function") {
        throw new Error(
          `babel-plugin-import style function is not supported in utoopack bundler`
        );
      }
      let transformRule = "{{ kebabCase member }}";
      if (camel2DashComponentName === false) {
        transformRule = "{{ member }}";
      }
      const skipDefaultConversion = typeof transformToDefaultImport === "undefined" ? false : !Boolean(transformToDefaultImport);
      acc[libraryName] = {
        transform: `${libraryName}/${libraryDirectory}/${transformRule}`,
        preventFullImport: false,
        skipDefaultConversion,
        style: typeof style === "boolean" ? "style" : style
      };
      return acc;
    },
    {}
  );
}
function getNormalizedAlias(alias, rootDir) {
  const newAlias = { ...alias };
  for (const [key, value] of Object.entries(newAlias)) {
    if (key.endsWith("/*") || value.endsWith("/*") || key.endsWith("/") || value.endsWith("/") || key.endsWith("$")) {
      continue;
    }
    const ext = (0, import_path.extname)(value);
    if (ext) {
      let isDirectory = false;
      const candidates = [.../* @__PURE__ */ new Set([value, (0, import_path.resolve)(rootDir, value)])];
      for (const candidate of candidates) {
        try {
          const stat = import_fs.default.statSync(candidate);
          if (stat.isDirectory()) {
            isDirectory = true;
            break;
          }
        } catch (e) {
        }
      }
      if (!isDirectory) continue;
    }
    newAlias[`${key}/*`] = `${value}/*`;
  }
  newAlias[`${rootDir}/*`] = `${rootDir}/*`;
  return newAlias;
}
function getNormalizedExternals(externals) {
  return Object.entries(externals || {}).reduce(
    (ret, [k, v]) => {
      if (Array.isArray(v)) {
        const [url, ...members] = v;
        const containsScript = url.startsWith("script");
        const script = url.replace("script ", "");
        if (containsScript) {
          ret[k] = {
            // ['antd', 'Button'] => `antd.Button`
            root: members.join("."),
            type: "script",
            // `script https://example.com/lib/script.js` => `https://example.com/lib/script.js`
            script
          };
        } else {
          ret[k] = {
            root: members.join("."),
            script
          };
        }
      } else if (typeof v === "string") {
        ret[k] = v.replace(/^window(\s+|\.)/, "");
      } else {
      }
      return ret;
    },
    {}
  );
}
function getSvgModuleRules(opts) {
  const { svgr, svgo = {} } = opts;
  if (!svgr) {
    return {};
  }
  return {
    module: {
      rules: {
        "*.svg": {
          loaders: [
            {
              loader: require.resolve("@umijs/bundler-webpack/dist/loader/svgr"),
              options: {
                svgoConfig: {
                  plugins: [
                    {
                      name: "preset-default",
                      params: {
                        overrides: {
                          removeTitle: false
                        }
                      }
                    },
                    "prefixIds"
                  ],
                  ...typeof svgo === "object" ? svgo : {}
                },
                ...svgr,
                svgo: !!svgo
              }
            },
            {
              loader: require.resolve("@umijs/bundler-webpack/compiled/url-loader")
            }
          ],
          as: "*.js"
        }
      }
    }
  };
}
function appendPostcssPlugin(postcssConfig, plugin) {
  const [pluginName, pluginOptions = {}] = plugin;
  if (postcssConfig == null) {
    return {
      plugins: {
        [pluginName]: pluginOptions
      }
    };
  }
  if (!import_utils.lodash.isPlainObject(postcssConfig)) {
    throw new Error(
      `Utoopack styles.postcss must be an object when extraPostCSSPlugins is used.`
    );
  }
  const existingPlugins = postcssConfig.plugins;
  if (existingPlugins != null && !import_utils.lodash.isPlainObject(existingPlugins)) {
    throw new Error(
      `Utoopack styles.postcss.plugins must be an object when extraPostCSSPlugins is used.`
    );
  }
  return {
    ...postcssConfig,
    plugins: {
      ...existingPlugins || {},
      [pluginName]: pluginOptions
    }
  };
}
function normalizeExtraPostcssPlugin(plugin) {
  if (typeof plugin === "string") {
    return [[plugin, {}]];
  }
  if (Array.isArray(plugin) && typeof plugin[0] === "string" && plugin.length <= 2) {
    return [[plugin[0], plugin[1] ?? {}]];
  }
  if (import_utils.lodash.isPlainObject(plugin)) {
    return Object.entries(plugin);
  }
  throw new Error(
    `Utoopack only supports JSON-serializable extraPostCSSPlugins entries. Please use plugin names or [pluginName, options] tuples instead.`
  );
}
function mergeExtraPostcssPlugins(postcssConfig, extraPlugins = []) {
  return extraPlugins.reduce((memo, plugin) => {
    return normalizeExtraPostcssPlugin(plugin).reduce(
      (ret, normalizedPlugin) => {
        return appendPostcssPlugin(ret, normalizedPlugin);
      },
      memo
    );
  }, postcssConfig);
}
function getUserUtoopackConfig(utoopackConfig = {}) {
  return import_utils.lodash.omit(utoopackConfig, ["root"]);
}
async function getProdUtooPackConfig(opts) {
  var _a;
  const webpackConfig = await (0, import_bundler_webpack.getConfig)({
    cwd: opts.cwd,
    rootDir: opts.rootDir,
    env: "production",
    entry: opts.entry,
    userConfig: opts.config,
    analyze: process.env.ANALYZE,
    babelPreset: opts.babelPreset,
    extraBabelPlugins: [
      ...opts.beforeBabelPlugins || [],
      ...opts.extraBabelPlugins || []
    ],
    extraBabelPresets: [
      ...opts.beforeBabelPresets || [],
      ...opts.extraBabelPresets || []
    ],
    extraBabelIncludes: opts.config.extraBabelIncludes,
    chainWebpack: opts.chainWebpack,
    modifyWebpackConfig: opts.modifyWebpackConfig,
    pkg: opts.pkg,
    disableCopy: opts.disableCopy
  });
  let utooBundlerOpts = (0, import_pack.compatOptionsFromWebpack)({
    ...import_utils.lodash.omit(webpackConfig, ["target", "module", "externals"]),
    webpackMode: true
  });
  const extraBabelPlugins = [
    ...opts.extraBabelPlugins || [],
    ...opts.config.extraBabelPlugins || []
  ];
  const modularizeImports = getModularizeImports(extraBabelPlugins);
  const emotion = extraBabelPlugins.some((p) => {
    return p === "@emotion" || p === "@emotion/babel-plugin";
  });
  const define = {};
  if (opts.config.define) {
    for (const key of Object.keys(opts.config.define)) {
      define[key] = normalizeDefineValue(opts.config.define[key]);
    }
  }
  if (process.env.SOCKET_SERVER) {
    define["process.env.SOCKET_SERVER"] = normalizeDefineValue(
      process.env.SOCKET_SERVER
    );
  }
  const {
    publicPath,
    runtimePublicPath,
    externals: userExternals,
    copy = [],
    svgr,
    svgo = {},
    inlineLimit
  } = opts.config;
  const userUtoopackConfig = getUserUtoopackConfig(opts.config.utoopack);
  utooBundlerOpts = {
    ...utooBundlerOpts,
    config: import_utils.lodash.merge(
      import_utils.lodash.omit(utooBundlerOpts.config, ["define"]),
      {
        output: {
          clean: opts.clean,
          publicPath: runtimePublicPath ? "runtime" : publicPath || "/",
          ...opts.disableCopy ? { copy: [] } : { copy: ["public"].concat(copy) }
        },
        optimization: {
          modularizeImports,
          concatenateModules: true
        },
        resolve: {
          alias: getNormalizedAlias(
            (_a = utooBundlerOpts.config.resolve) == null ? void 0 : _a.alias,
            opts.rootDir
          )
        },
        styles: {
          less: {
            modifyVars: opts.config.theme,
            javascriptEnabled: true,
            ...opts.config.lessLoader
          },
          // postcss: normalizedPostcssConfig,
          sass: opts.config.sassLoader ?? void 0,
          emotion
        },
        define,
        nodePolyfill: true,
        externals: getNormalizedExternals(userExternals),
        ...getSvgModuleRules({ svgr, svgo, inlineLimit })
      },
      userUtoopackConfig
    )
  };
  return utooBundlerOpts;
}
async function getDevUtooPackConfig(opts) {
  var _a;
  let webpackConfig = await (0, import_bundler_webpack.getConfig)({
    cwd: opts.cwd,
    rootDir: opts.rootDir,
    env: "development",
    entry: opts.entry,
    userConfig: opts.config,
    babelPreset: opts.babelPreset,
    extraBabelPlugins: [
      ...opts.beforeBabelPlugins || [],
      ...opts.extraBabelPlugins || []
    ],
    extraBabelPresets: [
      ...opts.beforeBabelPresets || [],
      ...opts.extraBabelPresets || []
    ],
    extraBabelIncludes: opts.config.extraBabelIncludes,
    chainWebpack: opts.chainWebpack,
    modifyWebpackConfig: opts.modifyWebpackConfig,
    // TO avoild bundler webpack add extra entry.
    hmr: false,
    analyze: process.env.ANALYZE
  });
  let utooBundlerOpts = (0, import_pack.compatOptionsFromWebpack)({
    ...import_utils.lodash.omit(webpackConfig, ["target", "module", "externals"]),
    webpackMode: true
  });
  const extraBabelPlugins = [
    ...opts.extraBabelPlugins || [],
    ...opts.config.extraBabelPlugins || []
  ];
  const modularizeImports = getModularizeImports(extraBabelPlugins);
  const emotion = extraBabelPlugins.some((p) => {
    return p === "@emotion" || p === "@emotion/babel-plugin";
  });
  const define = {};
  if (opts.config.define) {
    for (const key of Object.keys(opts.config.define)) {
      define[key] = normalizeDefineValue(opts.config.define[key]);
    }
  }
  if (process.env.SOCKET_SERVER) {
    define["process.env.SOCKET_SERVER"] = normalizeDefineValue(
      process.env.SOCKET_SERVER
    );
  }
  const {
    publicPath,
    runtimePublicPath,
    externals: userExternals,
    copy = [],
    svgr,
    svgo = {},
    inlineLimit
  } = opts.config;
  const userUtoopackConfig = getUserUtoopackConfig(opts.config.utoopack);
  utooBundlerOpts = {
    ...utooBundlerOpts,
    config: import_utils.lodash.merge(
      import_utils.lodash.omit(utooBundlerOpts.config, ["define"]),
      {
        output: {
          clean: opts.clean === void 0 ? true : opts.clean,
          publicPath: runtimePublicPath ? "runtime" : publicPath || "/",
          ...opts.disableCopy ? { copy: [] } : { copy: ["public"].concat(copy) }
        },
        resolve: {
          alias: getNormalizedAlias(
            (_a = utooBundlerOpts.config.resolve) == null ? void 0 : _a.alias,
            opts.rootDir
          )
        },
        optimization: {
          modularizeImports,
          // Disable the options under dev for utoopack
          // otherwise it will affect react-refresh
          removeUnusedExports: false,
          removeUnusedImports: false
        },
        styles: {
          less: {
            modifyVars: opts.config.theme,
            javascriptEnabled: true,
            ...opts.config.lessLoader
          },
          // postcss: normalizedPostcssPlugin,
          sass: opts.config.sassLoader ?? void 0,
          emotion
        },
        define,
        // dev enable persistent cache by default
        persistentCaching: true,
        nodePolyfill: true,
        externals: getNormalizedExternals(userExternals),
        ...getSvgModuleRules({ svgr, svgo, inlineLimit })
      },
      userUtoopackConfig
    ),
    watch: {
      enable: true
    },
    dev: true
  };
  return utooBundlerOpts;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getDevUtooPackConfig,
  getProdUtooPackConfig,
  mergeExtraPostcssPlugins
});
